### PR TITLE
added rate limit with exponential backoff and retry

### PR DESCRIPTION
**Pull request by jspaaks**
**Describe the changes made in this pull request**

- Added rate limit with exponential backoff and retry. Refs #93
- Removed sleep from livetests and clitests
- ~~Added `-a` and `--apikeys-filename` to cli. Refs #263~~ Now reading from env vars
- the functions with which to interact with github, gitlab, etc are in `requesting` and are organized as follows:

    ```
    .
    └── get_from_platform
        ├── get_from_github
        │   ├── get_from_github_no_auth
        │   │   ├── get_from_github_no_auth_api
        │   │   ├── get_from_github_no_auth_frontend
        │   │   └── get_from_github_no_auth_raw
        │   └── get_from_github_with_auth
        │       ├── get_from_github_with_auth_api
        │       ├── get_from_github_with_auth_frontend
        │       └── get_from_github_with_auth_raw
        └── get_from_gitlab
            ├── get_from_gitlab_no_auth
            │   ├── get_from_gitlab_no_auth_api
            │   ├── get_from_gitlab_no_auth_frontend
            │   └── get_from_gitlab_no_auth_raw
            └── get_from_gitlab_with_auth
                ├── get_from_gitlab_with_auth_api
                ├── get_from_gitlab_with_auth_frontend
                └── get_from_gitlab_with_auth_raw
    ```
    This gives the developer control over the level at which to use the rate limiting decorators, for example different rates apply for each platform, a given platform may have different rate limits for unauthenticated/authenticated, as well as different rates for different parts of their service.

**Instructions to review the pull request**

Try running the unit tests

```shell
pytest tests/
```

Now replace `get_calls()` in the decorator in `requesting/get_from_github_no_auth.py` with 60. Running the tests again will halt halfway through because of rate limits. It will continue very slowly due to exponential backoff and retry, but the retries happen too slow for the ratelimit of 60 per hour, therefore they all ultimately fail with a too-many-calls error.

~~cli now has a `-a | --apikeys-filename` option that can be used to point to a secrets file with api keys~~

Apikeys are read from env vars named `APIKEY_GITHUB` and `APIKEY_GITLAB,` e.g.:

```text
$ cat secrets.txt
APIKEY_GITHUB=<user who made the key>:<longish key with random characters>
APIKEY_GITLAB=<user who made the key>:<longish key with random characters>
```

```shell
source secrets.txt
howfairis https://github.com/fair-software/howfairis
```

**Open questions / problems**

1. ~~How to bypass the ratelimit decorators when using mocks~~ fixed it, but hides the next point about `Repo`
2. ~~`Repo` might need to be extended with ``apikeys`` arg as well, since it's making a request to `raw`.~~ No longer needed now that we're reading from env vars
3. What about instantiating `Repo` within `Checker.__init__`? `Checker`'s constructor would then need additional arguments `url`, `branch` and `path`, and could drop `repo`.
4. On GitHub and GitLab, `raw` seems to have its own limits, not covered by api key increased rate limit.
5. On GitLab, unauthenticated api calls return a ratelimit of 2000, not 500 like the documentation says. authenticated requests also have ratelimit 2000, this is consistent with the documentation

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,3 +9,5 @@
 
 docs/_build
 docs/apidocs
+
+secrets.yml
```
Comment (fdiblen) [inline .gitignore:13]:
```suggestion
```

File: howfairis/requesting/get_from_github.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from .get_from_github_no_auth import get_from_github_no_auth
+from .get_from_github_with_auth import get_from_github_with_auth
+
+
+def get_from_github(url, url_type, apikeys=None):
+
+    github_apikey_is_invalid = apikeys is not None and (apikeys.get("github-key") is None or
+                                                        apikeys.get("github-user") is None)
```
Comment (eriktks) [inline howfairis/requesting/get_from_github.py:8]: These lines are difficult to read

Comment (fdiblen) [inline howfairis/requesting/get_from_github.py:8]: How about something like:

if apikeys.get("github-key") and apikeys.get("github-user"):

File: howfairis/requesting/get_from_github.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+    if apikeys is None or github_apikey_is_invalid:
+        return get_from_github_no_auth(url, url_type)
```
Comment (fdiblen) [inline howfairis/requesting/get_from_github.py:11]:
```suggestion
    if apikeys is and apikeys.get("github-key") and apikeys.get("github-user"):
        return get_from_github_no_auth(url, url_type)
```

File: howfairis/requesting/get_from_github.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from .get_from_github_no_auth import get_from_github_no_auth
+from .get_from_github_with_auth import get_from_github_with_auth
+
+
+def get_from_github(url, url_type, apikeys=None):
+
+    github_apikey_is_invalid = apikeys is not None and (apikeys.get("github-key") is None or
+                                                        apikeys.get("github-user") is None)
+
+    if apikeys is None or github_apikey_is_invalid:
```
Comment (eriktks) [inline howfairis/requesting/get_from_github.py:10]:
```suggestion
    if not (apikeys and apikeys.get("github-key") and apikeys.get("github-user")):
```

Comment (eriktks) [inline howfairis/requesting/get_from_github.py:10]:
```suggestion
    if apikeys is None or apikeys.get("github-key") is None or apikeys.get("github-user") is None:
```

File: howfairis/requesting/get_from_gitlab.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from .get_from_gitlab_no_auth import get_from_gitlab_no_auth
+from .get_from_gitlab_with_auth import get_from_gitlab_with_auth
+
+
+def get_from_gitlab(url, url_type, apikeys=None):
+
+    if apikeys is None:
```
Comment (eriktks) [inline howfairis/requesting/get_from_gitlab.py:7]: This code is much shorter than the equivalent code for github. Is it complete?

File: howfairis/requesting/get_from_gitlab_with_auth_api.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+from ratelimit import limits
+
+
+# https://docs.gitlab.com/ee/user/gitlab_com/index.html#gitlabcom-specific-rate-limits
+@on_exception(expo, RateLimitException, max_tries=8)
+@limits(calls=2000, period=60)
+def get_from_gitlab_with_auth_api(url, apikeys):
+
+    headers = {
+        "Accept": "application/vnd.github.v3+json"
```
Comment (eriktks) [inline howfairis/requesting/get_from_gitlab_with_auth_api.py:14]: Probably github here should be changed

File: setup.py [hunk 7]
```diff
@@ -37,16 +38,18 @@
...
-        "ruamel.yaml>=0.16",
-        "requests>=2",
-        "voluptuous>=0.11"
+        "backoff == 1.*",
+        "beautifulsoup4 == 4.*",
+        "click == 7.*",
+        "colorama == 0.*",
+        "ratelimit == 2.*",
+        "ruamel.yaml == 0.16.*",
+        "requests == 2.*",
```
Comment (eriktks) [inline setup.py:51]: these two lines in alphabetical order

**Review by eriktks — APPROVED** [review]
We went through the code and added some minor comments. Well done!

**Review by fdiblen — APPROVED** [review]
Awesome sos! :ramen:
A few suggestions which we should create issues about:
- We need to update the documentation to reflect this work.
- It is a good idea to raise some exceptions, e.g. when the authentication fails.
- If for some reason we cannot use authenticantion and use unathenticated method, we should warn the user.
- Number of calls are hardcoded. We can consider making them configurable.

File: .gitignore (hunk 1 above)
Comment (jspaaks) [reply .gitignore:13]: done in upcoming commit

File: howfairis/requesting/get_from_gitlab.py (hunk 5 above)
Comment (jspaaks) [reply howfairis/requesting/get_from_gitlab.py:7]: fixed it same as github now

File: howfairis/requesting/get_from_gitlab_with_auth_api.py (hunk 6 above)
Comment (jspaaks) [reply howfairis/requesting/get_from_gitlab_with_auth_api.py:14]: changed to only json, also added headers to equivalent noauth gitlab case.

File: setup.py (hunk 7 above)
Comment (jspaaks) [reply setup.py:51]: fixed it

**Comment by jspaaks** [comment]
Spawned

1. #297 
2. #298
3. #299

the remaining bullet point from https://github.com/fair-software/howfairis/pull/286#pullrequestreview-598804116 is covered by #298

**Pull request merged**